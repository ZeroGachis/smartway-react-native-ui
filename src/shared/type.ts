export type WithTestID<Props> = Props & { testID?: string };
