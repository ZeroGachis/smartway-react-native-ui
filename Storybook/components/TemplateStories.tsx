import type { Meta, StoryObj } from '@storybook/react-native';
import React from 'react';
import { StyleSheet, View } from 'react-native';

// fake component must be replace by one from the library
const Component = (props: React.PropsWithChildren) => {
    const { children } = props;
    return <>{children}</>;
}; // must be removed

type ComponentProps = React.ComponentProps<typeof Component>;

export default {
    title: 'components/Component', // change it with your component name eg('components/AwesomeButton',)
    component: Component,
    //default value
    args: {},
    // add some control in argTypes
    // some example
    // https://github.com/storybookjs/react-native/blob/next/examples/expo-example/components/ControlExamples/ControlExample/ControlExample.stories.tsx
    argTypes: {},
    // use it to wrap your storie in composant
    decorators: [
        (Story) => {
            const styles = StyleSheet.create({
                container: { paddingTop: 16 },
            });
            return (
                <View style={styles.container}>
                    <Story />
                </View>
            );
        },
    ],
} as Meta<ComponentProps>;

type Story = StoryObj<ComponentProps>;

// first story use it with args to change load only props value
// you can also use render as
// render : (args) => <AwesomeButton {...args}>
export const Default: Story = {
    args: {},
};
Default.parameters = { noSafeArea: false };
