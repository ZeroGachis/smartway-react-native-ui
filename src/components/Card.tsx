import type { ReactNode } from "react";
import React from "react";
import { StyleProp, View, ViewStyle } from "react-native";
import { useTheme } from "smartway-react-native-ui";

interface Props {
    children?: ReactNode,
    style: ViewStyle
};

export const Card = ({children, style}: Props) => {
    const theme = useTheme();
    const cardStyle: StyleProp<ViewStyle> = {
        backgroundColor: theme.sw.colors.neutral[50],
        width: '100%',
        alignItems: 'center',
        borderRadius: 12,
        paddingTop: theme.sw.spacing.xl,
        paddingBottom: theme.sw.spacing.xl,
        paddingLeft: theme.sw.spacing.m,
        paddingRight: theme.sw.spacing.m,
        ...style
        // TODO: Implement shadows
    };

    return (
        <View style={cardStyle}>
            {children}
        </View>
    );
};
