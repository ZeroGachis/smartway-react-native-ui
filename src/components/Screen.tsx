import React from 'react';
import { SafeAreaView, ViewStyle, StatusBar, StyleSheet } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import { useTheme } from '../styles/themes';
import DeviceInfo from 'react-native-device-info';
import type { WithTestID } from 'src/shared/type';

type Props = WithTestID<{
    children?: React.ReactNode;
    style?: ViewStyle;
    statusBarColor?: string;
}>;

export const Screen = ({ children, style, testID, statusBarColor }: Props) => {
    const theme = useTheme();
    const insets = useSafeAreaInsets();
    const isTablet = DeviceInfo.isTablet();

    const styles = StyleSheet.create({
        screen: {
            flex: 1,
            backgroundColor: theme.sw.color.neutral[0],
            marginTop: insets.top,
            paddingLeft: isTablet ? theme.sw.spacing.s : 0,
            paddingRight: isTablet ? theme.sw.spacing.s : 0,
            ...style,
        },
    });

    return (
        <SafeAreaView style={styles.screen} testID={testID}>
            <StatusBar
                translucent
                backgroundColor={
                    statusBarColor
                        ? statusBarColor
                        : theme.sw.color.neutral[0]
                }
                barStyle='dark-content'
            />
            {children}
        </SafeAreaView>
    );
};
