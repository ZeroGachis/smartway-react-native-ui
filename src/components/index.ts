import { Button } from 'react-native-paper';
import Logo from './logos/logos';

export { Button, Logo };