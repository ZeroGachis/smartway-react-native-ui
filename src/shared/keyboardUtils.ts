import { useEffect } from 'react';
import { Keyboard, KeyboardEventListener, TextInput } from 'react-native';

export function useListenerOnKeyboardHiding(listener: KeyboardEventListener) {
    useEffect(() => {
        const hideSubscription = Keyboard.addListener(
            'keyboardDidHide',
            listener,
        );
        return () => {
            hideSubscription.remove();
        };
    }, [listener]);
}

export function hideKeyboard(ref: React.RefObject<TextInput>) {
    ref?.current?.blur();
}
