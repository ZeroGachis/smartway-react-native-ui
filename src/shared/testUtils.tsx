import React, { ReactElement } from 'react';
import {
    RenderOptions,
    render as rtlRender,
} from '@testing-library/react-native';
import { ThemeProvider } from '../styles/themes'; // Replace with the actual path to your ThemeProvider

const uiRender = (ui: ReactElement, options?: RenderOptions) => {
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
        <ThemeProvider>{children}</ThemeProvider>
    );

    return rtlRender(ui, { wrapper: Wrapper, ...options });
};

export * from '@testing-library/react-native';
export { uiRender as render };
