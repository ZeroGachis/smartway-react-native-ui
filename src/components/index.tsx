import { Button } from './buttons/Button';
import { IconButton } from './buttons/IconButton';
import { Logo } from './logos/Logo';
import { Screen } from './Screen';
import { Headline } from './typography/Headline';
import { Body } from './typography/Body';
import { Modal } from './modals/Modal';
import { TextField } from './fields/TextField';
import { Icon } from './icons/Icon';
import { TopAppBar } from './topAppBar/TopAppBar';
import { Card } from './card/Card';
import { Tag } from './tag/Tag';
import { Divider } from './divider/Divider';
import { LineChart } from './charts/LineChart';
import { Badge } from './badge/Badge';
import { DateField } from './dateSelector/DateField';
import { DateSelector } from './dateSelector/DateSelector';
import { Label } from './typography/Label';
import { FieldState, NumberField } from './numberField/NumberField';

export { default as Snackbar, useSnackbar } from './alert/Snackbar';
export { default as Banner, useBanner } from './alert/Banner';

export {
    Body,
    Button,
    IconButton,
    Modal,
    Headline,
    Icon,
    Logo,
    Label,
    Screen,
    TextField,
    TopAppBar,
    NumberField,
    FieldState,
    Card,
    Tag,
    Divider,
    LineChart,
    Badge,
    DateField,
    DateSelector,
};
