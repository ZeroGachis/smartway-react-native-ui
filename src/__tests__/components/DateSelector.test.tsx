import React from 'react';
import { DateSelector } from '../../components';
import { DeviceEventEmitter } from 'react-native';
import { render, userEvent, screen, act } from '../../shared/testUtils';

const mockedTestID = 'mockedTestID';
const mockOnChange = jest.fn();
let tree: ReturnType<typeof render>;

beforeEach(() => {
    tree = render(
        <DateSelector
            prefilled={new Date(2003, 1, 1)}
            onChange={mockOnChange}
            testID={mockedTestID}
        />,
    );
});

describe('MODULE | DateField', () => {
    it('component renders correctly with prefilled values in DD/MM/YY', () => {
        expect(tree.toJSON()).toMatchSnapshot();
    });

    it('should send date filled with missing fields', async () => {
        const user = userEvent.setup();

        const yearField = screen.getByTestId(mockedTestID + '/year');
        await user.type(yearField, '24');

        await act(() => {
            DeviceEventEmitter.emit('keyboardDidHide', {});
        });

        expect(mockOnChange).toHaveBeenCalledWith(new Date(2024, 1, 1));
    });

    it('should send date filled with all changed fields', async () => {
        const user = userEvent.setup();

        const dayField = screen.getByTestId(mockedTestID + '/day');
        const monthField = screen.getByTestId(mockedTestID + '/month');
        const yearField = screen.getByTestId(mockedTestID + '/year');

        await user.type(dayField, '12');
        await user.type(monthField, '06');
        await user.type(yearField, '24');

        await act(() => {
            DeviceEventEmitter.emit('keyboardDidHide', {});
        });

        expect(mockOnChange).toHaveBeenCalledWith(new Date(2024, 5, 12));
    });

    it('should prefix date fields with a 0 on blur', async () => {
        const user = userEvent.setup();

        const dayField = screen.getByTestId(mockedTestID + '/day');
        await user.type(dayField, '4');
        expect(dayField.props.value).toBe('04');

        const monthField = screen.getByTestId(mockedTestID + '/month');
        await user.type(monthField, '5');
        expect(monthField.props.value).toBe('05');

        const yearField = screen.getByTestId(mockedTestID + '/year');
        await user.type(yearField, '6');
        expect(yearField.props.value).toBe('06');
    });

    it('should display an error message', async () => {
        tree.rerender(
            <DateSelector
                prefilled={new Date(2003, 1, 1)}
                onChange={mockOnChange}
                testID={mockedTestID}
                errorMessage='an error'
            />,
        );

        expect(screen.getByText('an error')).toBeOnTheScreen();
    });
});
