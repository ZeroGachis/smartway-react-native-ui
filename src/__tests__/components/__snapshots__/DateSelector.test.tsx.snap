// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MODULE | DateField component renders correctly with prefilled values in DD/MM/YY 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    collapsable={false}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
        testID="mockedTestID"
      >
        <TextInput
          cursorColor="#18A586"
          keyboardType="numeric"
          maxLength={2}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="01"
          placeholderTextColor="#919EAB"
          selectTextOnFocus={true}
          selectionColor="#18A58629"
          style={
            {
              "backgroundColor": "#919EAB14",
              "borderColor": undefined,
              "borderRadius": 18,
              "borderWidth": 0,
              "color": "#919EAB",
              "fontFamily": "PublicSans-Bold",
              "fontSize": 20,
              "height": 48,
              "lineHeight": 23.5,
              "marginVertical": 0,
              "paddingVertical": 0,
              "width": 72,
            }
          }
          testID="mockedTestID/day"
          textAlign="center"
          value=""
        />
        <View
          style={
            {
              "alignItems": "center",
              "justifyContent": "center",
              "width": 26,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#212B36",
                  "fontFamily": "PublicSans-Bold",
                  "fontSize": 20,
                },
                {
                  "color": "#919EAB",
                  "fontSize": 20,
                },
              ]
            }
          >
            /
          </Text>
        </View>
        <TextInput
          cursorColor="#18A586"
          keyboardType="numeric"
          maxLength={2}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="02"
          placeholderTextColor="#919EAB"
          selectTextOnFocus={true}
          selectionColor="#18A58629"
          style={
            {
              "backgroundColor": "#919EAB14",
              "borderColor": undefined,
              "borderRadius": 18,
              "borderWidth": 0,
              "color": "#919EAB",
              "fontFamily": "PublicSans-Bold",
              "fontSize": 20,
              "height": 48,
              "lineHeight": 23.5,
              "marginVertical": 0,
              "paddingVertical": 0,
              "width": 72,
            }
          }
          testID="mockedTestID/month"
          textAlign="center"
          value=""
        />
        <View
          style={
            {
              "alignItems": "center",
              "justifyContent": "center",
              "width": 26,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#212B36",
                  "fontFamily": "PublicSans-Bold",
                  "fontSize": 20,
                },
                {
                  "color": "#919EAB",
                  "fontSize": 20,
                },
              ]
            }
          >
            /
          </Text>
        </View>
        <TextInput
          cursorColor="#18A586"
          keyboardType="numeric"
          maxLength={2}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="03"
          placeholderTextColor="#919EAB"
          selectTextOnFocus={true}
          selectionColor="#18A58629"
          style={
            {
              "backgroundColor": "#919EAB14",
              "borderColor": undefined,
              "borderRadius": 18,
              "borderWidth": 0,
              "color": "#919EAB",
              "fontFamily": "PublicSans-Bold",
              "fontSize": 20,
              "height": 48,
              "lineHeight": 23.5,
              "marginVertical": 0,
              "paddingVertical": 0,
              "width": 72,
            }
          }
          testID="mockedTestID/year"
          textAlign="center"
          value=""
        />
      </View>
    </View>
  </View>
</RNCSafeAreaProvider>
`;
